components/spotify/sources/common.js

```diff
@@ -0,0 +1,38 @@
+const spotify = require("../spotify.app.js");
+
+module.exports = {
+  dedupe: "unique",
+  props: {
+    spotify,
+    db: "$.service.db",
+    timer: {
+      type: "$.interface.timer",
+      default: {
+        intervalSeconds: 60 * 15,
+      },
+    },
+  },
+  methods: {
+    daysAgo(days) {
+      const daysAgo = new Date();
+      daysAgo.setDate(daysAgo.getDate() - days);
+      return daysAgo;
+    },
+    async *paginate(resourceFn, params={}) {
+      params.limit = 20;
+      params.offset = 0;
+      let done = false;
+      while (!done) {
+        const results = await resourceFn(params);
+        const { items } = results;
+        for (const item of items) {
+          yield item;
+        }
+        if (items.length < params.limit)
+            done = true;
+        params.offset += params.limit;
+        await new Promise(resolve => setTimeout(resolve, 500));
```

Comment:
Was there a specific reason you set this to half a second? I see that Spotify returns a `Retry-After` header that tells us the number of seconds to wait before sending another request. Could we use that, instead? https://developer.spotify.com/documentation/web-api/

Reply:
The Retry-After header is only sent if the response status is 429 (too many requests). I couldn't find what their usual rate limit is, but figured 2 requests per second would be a decent default to use for regular pagination. Would it be better to leave this part out and just let the retry method in the app file handle it if the rate limit is exceeded?

Reply:
I'd recommend checking for the presence of the `Retry-After` header here, and if it's absent / undefined, continue.  When a `Retry-After` mechanism is supported for an API like it is here, I'd recommend we use it, since a user can create multiple event sources that would trigger the rate limit even at a 2/sec rate for an individual source.